runtime/src/epoch_stakes.rs

```diff
@@ -134,13 +138,58 @@ impl EpochStakes {
 #[derive(Debug, Clone, PartialEq, Serialize, Deserialize)]
 pub enum VersionedEpochStakes {
     Current {
-        stakes: Stakes<Stake>,
+        stakes: StakesSerdeWrapper,
         total_stake: u64,
         node_id_to_vote_accounts: Arc<NodeIdToVoteAccounts>,
         epoch_authorized_voters: Arc<EpochAuthorizedVoters>,
     },
 }
 
+/// Wrapper struct with custom serialization to support serializing
+/// `Stakes<StakeAccount>` as `Stakes<Stake>` without doing a full deep clone of
+/// the stake data. Serialization works by building a `Stakes<&Stake>` map which
+/// borrows `&Stake` from `StakeAccount` entries in `Stakes<StakeAccount>`. Note
+/// that `Stakes<&Stake>` still copies `Pubkey` keys so the `Stakes<&Stake>`
+/// data structure still allocates a fair amount of memory but the memory only
+/// remains allocated during serialization.
+#[cfg_attr(feature = "frozen-abi", derive(AbiExample, AbiEnumVisitor))]
+#[derive(Debug, Clone, PartialEq)]
+pub enum StakesSerdeWrapper {
+    Stake(Stakes<Stake>),
+    Account(Stakes<StakeAccount<Delegation>>),
+}
```

Comment:
This looks like _almost_ equal to `StakesEnum`.
The wrapping struct is also `VersionedEpochStakes`.
So possibly could just use `StakesEnum` with a custom `#[serde(with = "...")]`.

It is true that the new type adds more type safety, but that will be caught in serialize and looks like `StakesEnum::Delegations` is planned to be deprecated anyways:
https://github.com/anza-xyz/agave/blob/83bb29e5b/runtime/src/epoch_stakes.rs#L165-L167

Reply:
Yeah but we can't serialize `Stakes<Delegation>` as `Stakes<Stake>` because it's missing a field

Reply:
right, that is what I meant by "will be caught in serialize", (e.g. return an error or `debug_assert` there).

anyways, not insisting on this approach, as the new type does add some type safety; so up to you.

Reply:
Ah got it, wasn't sure exactly what type safety you were referring to. Yeah I think the type safety is nice so I prefer to stick with this approach.